src/plonkish/backend/halo2.rs

```diff
@@ -357,160 +388,248 @@ pub fn to_halo2_advice<F: PrimeField>(
+        let mut fixed = vec![vec![F::default(); n]; fixed_count];
+
+        let mut copies = vec![];
+        for subcircuit in self.sub_circuits.iter() {
+            for (column, values) in subcircuit.circuit.fixed_assignments.iter() {
+                let column = fixed_columns.get(&column.uuid()).unwrap();
+
+                for (offset, value) in values.iter().enumerate() {
+                    fixed[column.index][offset] = *value;
+                }
+            }
+            subcircuit.collect_permutations(cs, &mut copies);
+        }
 
-        Ok(())
+        Preprocessing {
+            permutation: halo2_middleware::permutation::AssemblyMid { copies },
+            fixed,
+        }
     }
+}
 
-    fn configure(_: &mut ConstraintSystem<F>) -> Self::Config {
-        unreachable!()
-    }
+pub fn halo2_verify(
+    proof: Vec<u8>,
+    params: ParamsKZG<Bn256>,
+    vk: VerifyingKey<G1Affine>,
+    instance: Vec<Vec<Fr>>,
+) -> Result<(), ErrorBack> {
+    // Verify
+    let start = Instant::now();
+    println!("Verifying...");
+    let mut verifier_transcript =
+        Blake2bRead::<_, G1Affine, Challenge255<_>>::init(proof.as_slice());
+    let verifier_params = params.verifier_params();
+    let strategy = SingleStrategy::new(&verifier_params);
+
+    let result = verify_proof_single::<KZGCommitmentScheme<Bn256>, VerifierSHPLONK<Bn256>, _, _, _>(
+        &verifier_params,
+        &vk,
+        strategy,
+        instance,
+        &mut verifier_transcript,
+    );
+    println!("Verify: {:?}", start.elapsed());
+    result
 }
 
-#[derive(Debug, Default)]
-pub struct ChiquitoHalo2SuperCircuit<F: PrimeField + From<u64>> {
-    sub_circuits: Vec<ChiquitoHalo2<F>>,
-    witness: SuperAssignments<F>,
+pub struct Halo2Setup {
+    cs: ConstraintSystemMid<Fr>,
+    pub params: ParamsKZG<Bn256>,
+    pub vk: VerifyingKey<G1Affine>,
+    pk: ProvingKey<G1Affine>,
+    pub circuits: Vec<ChiquitoHalo2<Fr>>,
 }
 
-impl<F: PrimeField + From<u64> + Hash> ChiquitoHalo2SuperCircuit<F> {
-    pub fn new(sub_circuits: Vec<ChiquitoHalo2<F>>, witness: SuperAssignments<F>) -> Self {
+impl Halo2Setup {
+    pub fn new(
+        cs: ConstraintSystemMid<Fr>,
+        params: ParamsKZG<Bn256>,
+        vk: VerifyingKey<G1Affine>,
+        pk: ProvingKey<G1Affine>,
+        circuits: Vec<ChiquitoHalo2<Fr>>,
+    ) -> Self {
         Self {
-            sub_circuits,
-            witness,
+            cs,
+            params,
+            vk,
+            pk,
+            circuits,
         }
     }
 
-    pub fn instance(&self) -> Vec<Vec<F>> {
-        let mut result = Vec::new();
-
-        for sub_circuit in &self.sub_circuits {
-            if !sub_circuit.circuit.exposed.is_empty() {
-                let instance_values = sub_circuit.circuit.instance(
-                    self.witness
-                        .get(&sub_circuit.ir_id)
+    pub fn generate_proof(
+        &self,
+        mut rng: BlockRng<DummyRng>,
```

Comment:
@alxkzmn we have already this rng when we are doing the setup? could this one be different that the one we pass to setup?

Also, why a hashmap of assignments instead of an assignment? For the super circuit? What do you think is the best interface for this?

Reply:
Perhaps there could be Halo2Setup and Halo2SetupSuperCircuit. with the first one accepting Assignments<Fr> as witness and the second one accepting the result of doing the mapping?